### PR TITLE
Add RayService Manifests for Stable Diffusion TPU Examples

### PR DESCRIPTION
**Pull request by ryanaoleary**
## Why are these changes needed?

Example RayService manifests using TPUs for documentation added in this PR: https://github.com/ray-project/ray/pull/45924.
Once https://github.com/ray-project/serve_config_examples/pull/8 is merged I'll change the working_dir field for both RayServices (currently they point to my fork).

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+kind: RayService
+metadata:
+  name: s-d-tpu
+spec:
+  serveConfigV2: |
+    applications:
+      - name: stable_diffusion
+        import_path: stable_diffusion.stable_diffusion_tpu:deployment
+        runtime_env:
+          working_dir: "https://github.com/ryanaoleary/serve_config_examples/archive/refs/heads/tpu-example.zip"
```
Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:11]: Leaving a note here so we don't forget to revert the fork back to kuberay instead of ryanaoleary 

I'm assuming this was intentional for testing purposes

Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:11]: As a side note, other samples in the kuberay repo have been putting source files within the same directory e.g.: https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/pytorch-resnet-image-classifier/fine-tune-pytorch-resnet-image-classifier.py 

Wonder if we should do the same here instead of putting them in the serve_config_example repo in https://github.com/ray-project/serve_config_examples/pull/8

@kevin85421 what do you think? The benefit to putting it in the kuberay repo is that the whole example is contained in the same repo

Comment (kevin85421) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml:11]: I'm fine with both options. The advantage of adding the YAML in the KubeRay repository is that we have complete control and can prevent unexpected updates to `serve_config_example`. However, the downside is that the KubeRay repository is much larger than `serve_config_example`, which could increase startup time.

Currently, the KubeRay repository is 187 MB. It increases the startup, but maybe not that much (depending on the user's network speed).

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            - pillow
+            - diffusers==0.7.2
+            - transformers==4.24.0
+            - flax
+            - ml_dtypes==0.2.0
+            - jax[tpu]==0.4.11
+            - -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
+            - fastapi
+            - pillow
+            - ray[serve,default]==2.9.0
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:24]: Why do we need to install Ray Serve here?

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          pip:
+            - pydantic<2
+            - google-api-python-client
+            - pillow
+            - diffusers==0.7.2
+            - transformers==4.24.0
+            - flax
+            - ml_dtypes==0.2.0
+            - jax[tpu]==0.4.11
+            - -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:21]: I haven't seen `-f` before. I am not sure whether this works or not.

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            - ray[serve,default]==2.9.0
+  rayClusterConfig:
+    rayVersion: '2.9.0' # Should match the Ray version in the image of the containers
+    ######################headGroupSpecs#################################
+    # Ray head pod template.
+    headGroupSpec:
+      # The `rayStartParams` are used to configure the `ray start` command.
+      # See https://github.com/ray-project/kuberay/blob/master/docs/guidance/rayStartParams.md for the default settings of `rayStartParams` in KubeRay.
+      # See https://docs.ray.io/en/latest/cluster/cli.html#ray-start for all available options in `rayStartParams`.
+      rayStartParams:
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:33]: `rayStartParams: {}`

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+              requests:
+                cpu: "1"
+                ephemeral-storage: 20Gi
+                google.com/tpu: "4"
+                memory: 80G
+          nodeSelector:
+            cloud.google.com/gke-tpu-accelerator: tpu-v4-podslice
+            cloud.google.com/gke-tpu-topology: 2x2x2
+        metadata:
+          labels:
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:92]: Why do we need these labels?

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            - pillow
+            - diffusers==0.7.2
+            - transformers==4.24.0
+            - flax
+            - ml_dtypes==0.2.0
+            - jax[tpu]==0.4.11
+            - -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
+            - fastapi
+            - pillow
+            - ray[serve,default]==2.9.0
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.tpu-single-host.yaml:24]: same

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  rayClusterConfig:
+    rayVersion: '2.9.0' # Should match the Ray version in the image of the containers
+    ######################headGroupSpecs#################################
+    # Ray head pod template.
+    headGroupSpec:
+      # The `rayStartParams` are used to configure the `ray start` command.
+      # See https://github.com/ray-project/kuberay/blob/master/docs/guidance/rayStartParams.md for the default settings of `rayStartParams` in KubeRay.
+      # See https://docs.ray.io/en/latest/cluster/cli.html#ray-start for all available options in `rayStartParams`.
+      rayStartParams:
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.tpu-single-host.yaml:33]: same

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+              requests:
+                cpu: "1"
+                ephemeral-storage: 20Gi
+                google.com/tpu: "4"
+                memory: 80G
+          nodeSelector:
+            cloud.google.com/gke-tpu-accelerator: tpu-v4-podslice
+            cloud.google.com/gke-tpu-topology: 2x2x1
+        metadata:
+          labels:
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.tpu-single-host.yaml:92]: same

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                google.com/tpu: "4"
+                memory: 80G
+              requests:
+                cpu: "1"
+                ephemeral-storage: 20Gi
+                google.com/tpu: "4"
+                memory: 80G
+          nodeSelector:
+            cloud.google.com/gke-tpu-accelerator: tpu-v4-podslice
+            cloud.google.com/gke-tpu-topology: 2x2x1
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.tpu-single-host.yaml:90]: How can I determine whether a TPU PodSlice is single-host or multi-host based on `gke-tpu-topology`?

Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-single-host.yaml:90]: This page has information about TPU topologies and whether they're single or multi host: https://cloud.google.com/kubernetes-engine/docs/concepts/tpus#autopilot-tpu-topology. In general, you can determine if a PodSlice is multi-host by taking the product of the topology (i.e. the total number of chips) and dividing it by the `google.com/tpu` value in the k8s_resource_limits (i.e. the number of TPU chips per Pod or VM host). In this example, the topology has 4 chips total and `google.com/tpu`=4, so it is single host.

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 3 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml:21]: It just specifies the link to search, from the runtime_env_setup.log I can see that it successfully runts it.

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 5 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml:92]: I don't think these are needed actually, I'll remove them.

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 1 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml:11]: I added the python code in `serve_config_example` since that's where the existing [stable diffusion example](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray-service.stable-diffusion.yaml) put it. I can move it if that's preferable, but `serve_config_example` doesn't seem to be changed much so I think it should be fine to leave it there.

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml (hunk 9 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-single-host.yaml:88]: Added a link for this in [3b44939](https://github.com/ray-project/kuberay/pull/2198/commits/3b449390808f87876c499aacb31dac657b903191).

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml (hunk 8 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-single-host.yaml]: Done in [3b44939](https://github.com/ray-project/kuberay/pull/2198/commits/3b449390808f87876c499aacb31dac657b903191).

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml (hunk 7 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-single-host.yaml]: [3b44939](https://github.com/ray-project/kuberay/pull/2198/commits/3b449390808f87876c499aacb31dac657b903191)

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml (hunk 6 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-single-host.yaml]: [3b44939](https://github.com/ray-project/kuberay/pull/2198/commits/3b449390808f87876c499aacb31dac657b903191)

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 4 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml]: Done in [3b44939](https://github.com/ray-project/kuberay/pull/2198/commits/3b449390808f87876c499aacb31dac657b903191).

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 2 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml]: I don't think it's needed actually, removed in[3b44939](https://github.com/ray-project/kuberay/pull/2198/commits/3b449390808f87876c499aacb31dac657b903191).

**Review by kevin85421 — COMMENTED** [review]
@andrewsykim I don't have multi-host TPU from my side. I will merge the PR if you approve it. Thanks!

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            ports:
+            - containerPort: 6379
+              name: gcs
+            - containerPort: 8265
+              name: dashboard
+            - containerPort: 10001
+              name: client
+            - containerPort: 8000
+              name: serve
+            volumeMounts:
```
Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:47]: I think we can remove this volume mount

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+              - mountPath: /tmp/ray
+                name: ray-logs
+            resources:
+              limits:
+                cpu: "2"
+                memory: "8G"
+              requests:
+                cpu: "2"
+                memory: "8G"
+          volumes:
```
Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:57]: Remove volumes

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 12]
```diff
@@ -0,0 +1,88 @@
+apiVersion: ray.io/v1
+kind: RayService
+metadata:
+  name: s-d-tpu
```
Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:4]: stable-diffusion-tpu

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            ports:
+            - containerPort: 6379
+              name: gcs
+            - containerPort: 8265
+              name: dashboard
+            - containerPort: 10001
+              name: client
+            - containerPort: 8000
+              name: serve
+            volumeMounts:
```
Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.tpu-single-host.yaml:47]: Remove volume mount

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+              - mountPath: /tmp/ray
+                name: ray-logs
+            resources:
+              limits:
+                cpu: "2"
+                memory: "8G"
+              requests:
+                cpu: "2"
+                memory: "8G"
+          volumes:
```
Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.tpu-single-host.yaml:57]: Remove volumes

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 12 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml]: [80fa07a](https://github.com/ray-project/kuberay/pull/2198/commits/80fa07a4754fdde5b510165b1b18a6440a1d1d50)

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml (hunk 14 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-single-host.yaml]: Done in f88d93b

File: ray-operator/config/samples/ray-service.tpu-single-host.yaml (hunk 13 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-single-host.yaml]: f88d93b

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 11 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml]: f88d93b

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 10 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml]: Done in f88d93b

**Comment by ryanaoleary** [comment]
> @andrewsykim I don't have multi-host TPU from my side. I will merge the PR if you approve it. Thanks!

I think it's important to merge https://github.com/ray-project/serve_config_examples/pull/8 first so I can change the working_dir to point to the kuberay/master branch rather than my fork.

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml (hunk 1 above)
Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml]: Done in [98c5312](https://github.com/ray-project/kuberay/pull/2198/commits/98c531208f3e4d2b2bc6901862a026d10376568b).

File: ray-operator/config/samples/ray-service.tpu-multi-host.yaml [hunk 15]
```diff
@@ -72,15 +72,15 @@ spec:
             image: rayproject/ray:2.9.0-py310
             resources:
               limits:
-                cpu: "1"
+                cpu: "100"
```
Comment (andrewsykim) [inline ray-operator/config/samples/ray-service.tpu-multi-host.yaml:75]: typo to use 100 CPUs?

Comment (ryanaoleary) [reply ray-operator/config/samples/ray-service.tpu-multi-host.yaml]: [0ccb431](https://github.com/ray-project/kuberay/pull/2198/commits/0ccb4318d630f2d71243de8e3725f29562c5e7a8)

**Comment by ryanaoleary** [comment]
Update: removing the multi-host TPU RayService example from this PR, to be added later as part of a separate guide.

**Comment by kevin85421** [comment]
cc @andrewsykim is this PR ready to merge?

**Review by andrewsykim — APPROVED** [review]
LGTM

**Review by kevin85421 — APPROVED** [review]
Nice!
